stan/math/gpu/constants.hpp

```diff
@@ -0,0 +1,17 @@
+#ifndef STAN_MATH_GPU_CONSTANTS_HPP
+#define STAN_MATH_GPU_CONSTANTS_HPP
+#ifdef STAN_OPENCL
+namespace stan {
+namespace math {
+namespace gpu {
+const int Lower = 0;
```

Comment:
Initially these were enums, but after the discussion ([here](https://github.com/stan-dev/math/pull/916#issuecomment-404149485)) with Daniel and Bob, the decision was made to change this to `const int`.

Reply:
Ah, damn. The thing that didn't make it out in that conversation is that C++11 scoped enums can be template parameters 😂

Reply:
Does this mean we could use enums? I would like that very much.

Reply:
@bob-carpenter do you want to weigh in on the use of [scoped enums](https://en.cppreference.com/w/cpp/language/enum#Scoped_enumerations) in template parameters? It seems to be [somewhat standard C++](https://stackoverflow.com/questions/9116267/using-an-enumeration-as-a-template-parameter).

Reply:
@rok-cesnovar I think it should be fine; they're a part of C++ people generally like, though there are times when they are inappropriate or require more code I think we'd end up being fine here. Let's see if Bob has time today to give a quick double check. And just to clarify - it would be okay to merge this PR without changing these back to enums, since they already exist in the code base as ints and they already passed review that way. So this is an optional 2nd (3rd?) refactoring, haha.

Reply:
I was thinking making something along these lines:
```
enum class TriangularView { Lower = 0, Upper = 1, Entire = 2 };
enum class TriangularMap { UpperToLower = 0, LowerToUpper = 1 };
```
Do we still need the `gpu` namespace in this case or should we add a gpu/cl suffix or prefix?

Reply:
I would say these can go wherever the rest of the code is going. But if these are just internal implementation details (they are, right?) then they can go in the `stan::math::internal` namespace, which is where we hide those.

Reply:
If someone would use Stan Math and would want to use the GPU functions like `zeros`, `copy_triangular`, etc. he would need to specify the template parameter, so I guess this are not internal, idk.

Reply:
idt these are internal since users who want to write new functions will want to access them.

Reply:
Oh my bad, you guys are right. So either a gpu namespace (I'm not sure of the origin story there) or the `stan::math` one with a clear name imo

Reply:
I think the origin was due to not knowing about scoped enums. It didn't seem right to drop all of these at the Stan math namespace level.

Reply:
I think probably just in `stan::math` then - based on our overloading-to-get-rid-of-unecessary-transfers discussion, we'll need gpu functions in `stan::math`, so these might as well be in the same namespace.